### ISSUE TITLE
Add Markdown table support

### ISSUE THREAD

**Issue by Isvvc** [nextcloud/text#114]
**Is your feature request related to a problem? Please describe.**
Yes. Documents with Markdown tables do not have their tables rendered.

**Describe the solution you'd like**
Standard Markdown table rendering.

**Describe alternatives you've considered**
As developing a UI for making tables could be difficult, simply using the standard Markdown to HTML table conversion would be enough.

**Comment by juliushaertl** [comment]
Yes, we should look into implementing support for https://github.github.com/gfm/#tables-extension-

**Comment by jancborchardt** [comment]
@Isvvc is this something you would like to help contribute? 😊 The Text app is of course at the first version, and any contribution is welcome – we’ll help you get set up if you need anything.

**Comment by Isvvc** [comment]
@jancborchardt I could definitely make an attempt but am not sure exactly where to start. I have experience with text manipulation but none with Nextcloud apps or Vue.js.

**Comment by jancborchardt** [comment]
@Isvvc cool! Maybe @juliushaertl or someone from @nextcloud/vue has some pointers to help. :)

**Comment by ChristophWurst** [comment]
For Vue I suggest to read the Vue docs and then some code from Nextcloud apps. This will make you understand how we use it here.

---

> someone from @nextcloud/vue has some pointers to help. :)

I would kindly ask to not abuse the team for stuff like this. This just floods everyone's notifications. The maintainers of this app should be able to help.

Unsubscribed. Please @mention me directly if you need any specific help.

**Comment by juliushaertl** [comment]
Regarding the integration tables, I would recomment to first have a look at the tiptap table nodes to see what they support and how/if we can represent that in github flavored markdown. Since most likely not all will be supported, we need to create our own table nodes or inherit from the tiptap ones and add them to https://github.com/nextcloud/text/blob/master/src/EditorFactory.js#L45-L59

Here are some rough steps with code hints:
- Make sure markdownit actually renders the tables: https://github.com/nextcloud/text/blob/master/src/EditorFactory.js#L73-L74
- Make sure node definitions properly parse the rendered output to the table node types
- Implement tiptap to markdown conversion in custom nodes https://github.com/nextcloud/text/blob/master/src/marks/index.js#L46-L75

**Comment by elpraga** [comment]
Is there a roadmap as approximately when this functionality could be implemented?

**Comment by juliushaertl** [comment]
No, there is no estimation that we can give. Of course contributions are always welcome.

**Comment by elpraga** [comment]
@juliushaertl I wish i were able to contribute with the code.  I'm an English teacher,  not a coder.  Even though I did have a look at tiptap tables (they work great BTW), I was not able to make them work locally. 

I'm grateful that nextcloud exists and gives us the freedom to host files at home instead of having to rely on a big corporation. I do file bug reports and try to contribute with suggestions,  but,  unfortunately,  I cannot contribute with code. 

Thank you for your work,  and if tables in the best app could be implemented,  it would be much appreciated.

**Comment by juliushaertl** [comment]
@elpraga Thank you very much the kind words. Keep your contributions coming :wink:

**Comment by jonathandeclan** [comment]
Since this uses tiptap, I don't understand why we don't have tables
https://tiptap.scrumpy.io/tables

**Comment by guddl** [comment]
> Since this uses tiptap, I don't understand why we don't have tables
> https://tiptap.scrumpy.io/tables

Yes please :-)

**Comment by juliushaertl** [comment]
@jonathandeclan @guddl Please read https://github.com/nextcloud/text/issues/114#issuecomment-507180208 and use [Github Reactions](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) feature instead of commenting that also need a specific feature. That way we can see your support without notifying all users subscribed to a given issue without adding anything new to the discussion.

TLDR: As tiptap doesn't have markdown support out of the box, additional implementation work is required. Contributions are always welcome.

**Comment by jakobroehrl** [comment]
It would be awesome if the tables could be sorted and are filterable :)

**Comment by markuskramerIgitt** [comment]
I happily used markdown tables and assumed there were a standard feature of Nextcloud.


After an upgrade from Nextcloud 16 to 17, my markdown tables are no longer rendered as html table.


Can someone confirm that Nextcloud 16 renders markdown tables?

**Comment by davidbe** [comment]
It would be nice to hav an implementation like this: https://blog.documentnode.io/when-markdown-meets-tables

**Comment by fajabird** [comment]
What about an option to switch to raw markdown editing so any formatting supported by the rendering engine can be used even when not supported by the text-editing UI. This of course would require the support of extended markdown syntax by the engine.

I think as this editor used markdown in the backend there should always be an option to switch to RAW mode and supported elements should be rendered as expected. E.g. when I copy paste an markdown file to my nextcloud foldert and open it with Text it should display all elements correctly, not only those that are supported by the UI. For the time beeing those not supported by the UI could be marked so its obvious that those can only be edited from the RAW mode.

**Comment by juliushaertl** [comment]
@fajabird See #123

**Comment by asoltesz** [comment]
This is a pretty big blocker for any realistic use of PicoCMS with NextCloud since any moderately complex page will contain tables.

**Comment by Daksol** [comment]
Version of Markdown in Mattermost would be useful, it has Tables.

**Comment by aharish** [comment]
Just stumbled across this. I will pick this up over the weekend and see if I can help contribute something. Never written in Due before, but I have a solid understanding of other JS frameworks, so it mustn't be that hard to reorient myself.

I have worked with Markdownit on other platforms previously, so I think I should be able to contribute a very simple working version of table rendering support on the fancy pants editor.

**Comment by juliushaertl** [comment]
@aharish Feel free to ping me if any questions come up :)

**Comment by mikini** [comment]
For people looking for an existing work-around for table rendering in Nextcloud, the possibility exists to install the [Markdown Editor app](https://apps.nextcloud.com/apps/files_markdown) which includes a combined plain text editor and Markdown render with table support ([also based on markdown-it](https://github.com/icewind1991/files_markdown/blob/master/package.json#L5-L12) which supposedly supports [GFM tables by default](https://github.com/markdown-it/markdown-it#syntax-extensions)).

**Comment by b3** [comment]
@mikini This is not the same use case: there is no current multi-user support in Markdown Editor or am I missing something?

**Comment by mikini** [comment]
As I understand Isvvc's request, it primarily regards rendering of a markdown formatted file. Editing is not mentioned in the original issue text at all and Markdown Editor is able to do rendering of tables.

But you are right about collaborative editing not working using Mardown Editor. Just tested on my NC 20.0.7 instance with Markdown Edtior 2.3.3 and Plain text editor 2.14.0. Using the Markdown Editor requires reload of the file to pick up changes to the document, and will result in conflict and lost edits if two users are editing simultaneously. And if mixing editing with text (this application) it seems to be even more messy, so that is definitely not a work-around to recommend on multi-user instances.

**Comment by FrViPofm** [comment]
Multi-editor solution could come in a second time.
I think the main needs are for an editor :

1. that don't break the code (tables and yaml)
2. that can edit table (and yaml) in raw mode
3. that work on table in wysiwyg mode in single-user mode
4. that work on multi-user mode
5. that make the coffee and have a tour outside with the pet.

**Comment by johnwry** [comment]
Is there any progress on this issue? I can't make tables render

**Comment by johnwry** [comment]
I found my issue. I was using Nextcloud's default Text app and I should have been using the Plain Text editor app instead along with Markdown editor.

**Comment by MADXhh** [comment]
I cannot render tables.:-(
Which app exactly do I have to use/install?
And which table syntax can I use?

I just clicked on the plus sign, created a new file Test.md and added the following:

| Item | Price | # In stock |
|--------------|-----------|------------|
| Juicy Apples | 1.99 | *7* |
| Bananas | **1.89** | 5234 |

Unfortunately, the table is not displayed/rendered as a table, only the plain text.
NC version 22

Many thanks in advance for your support

**Comment by mejo-** [comment]
Hey @MADXhh, thanks for your feedback. Unfortunately, rendering tables is not supported by the Nextcloud Text app yet. It's a high-priority issue though and we most likely start working on it soon :blush:

**Comment by gravaigu** [comment]
As a simple workaround, I use the code block syntax, like this:

```
| Révision au | Montant en Euros |
|:-----------:|:----------------:|
|  01/01/2021 |     641,38 €     |
|  01/01/2020 |     641,75 €     |
|  01/01/2019 |     636,93 €     |
|  01/01/2018 |     626,62 €     |
|  01/01/2017 |     619,76 €     |
|  01/01/2016 |     616,36 €     |
```

(enclose the table with ``` before and after).

**Comment by ktprograms** [comment]
Hi, I think I may have thought of a way to edit tables in a html `textarea`, inspired by emacs org mode tables.

You can try it out by opening https://ktprograms.codeberg.page/table_editor_html/@master/, the instructions are at the top of that page.

You can leave feedback on the [issues page for my repo](https://codeberg.org/ktprograms/table_editor_html/issues), or on this issue thread.

Thanks!

**Comment by aharish** [comment]
Hi @ktprograms, this issue is not about typing down the table itself, it is about the Markdown renderer that Text uses not rendering those Markdown tables in HTML. While your provided solution is a good enhancement, it is not relevant to this issue. I would recommend that you create a new issue and post this there. Even better would be if you could submit a PR with this implemented on the Text app. It would be a nice enhancement to have!

Happy New Year and best wishes!

**Comment by ktprograms** [comment]
@aharish I see, thanks! I'll see about creating a PR but the current code is very minimal and needs more work.

The idea I was having was to use this in Nextcloud Notes, where (since the checkboxes are already only rendered in edit mode), the table wouldn't actually need any rendering since the Markdown table format is readable as-is.

**Comment by max-nextcloud** [comment]
I had a brief look at tiptap tables, markdown table syntax (gfm). Gfm tables are really lightweight. They do not allow for other nodes inside the tables - so no lists or even multiple paragraphs in the table cells. This will make tables somewhat less powerful but also easier to implement.

I think a meaningful first step would be rendering existing tables in Markdown, showing them in the editor and allowing to change the content. This would require:

- [ ] Adding a table extension to markdownit to parse the tables in markdown to html
- [ ] Adding a table extension to tiptap and limiting it to only allow the table structures markdown permits:
  - [ ] only one (optional) table header on top
  - [ ] no nested nodes inside table cells (content: marks or so for the table cell node?)
  - [ ] Add a `toMarkdown` function to the different nodes that make up a table.

If we add all that we would be able to:
* display existing tables in markdown
* handle copy and paste of html tables (removing all nested marks)
* allow editing of table content (it's simply another kind of node in tiptap)
* allow removing entire tables etc.

It would not include menu bar items to create a table, add / delete rows and columns, change alignment etc.
I think this would significantly reduce the need for UI discussions around creating / editing table structures.

**Comment by juliushaertl** [comment]
Sounds good, we can look into the table inserting and adding/removing columns in a second step.

**Comment by mejo-** [comment]
Thanks for looking into this @max-nextcloud, sounds really promising :+1:

I also think that keeping tables simple makes sense. There's no way to represent tables with multiline cells or other complex formatting inside the cells in markdown anyway, so it makes sense to me to restrict this to single-line tables and only marks (no futher nodes) allowed inside them.

We definitely need menu bar tools for creating/modifiying tables in my opinion, but that certainly can happen in a second step.

**Comment by ajoana** [comment]
> Multi-editor solution could come in a second time. I think the main needs are for an editor :
> 
> 1. that don't break the code (tables and yaml)
> 2. that can edit table (and yaml) in raw mode and preview it
> 3. that work on table in wysiwyg mode in single-user mode
> 4. that work on multi-user mode
> 5. that make the coffee and have a tour outside with the pet.

As a normal user, what could I do to stimulate progress on this issue? Could I sponser someone to get it done quickly? To me, the limitations of the default text app is a MAIN show stopper to better use and recommend NC.
I love collectives (text app mandatory) and wish to publish parts of the collectives using pico cms or other online md publishing tools.
1. this would be an urgent request (see pico cms and other issues)
2. could be a second step after 1.
3. this is an urgent request seeing this open here more than 1,5 years.
4. could be a second step after 3 - I don't see this important for our group as focus is on collectives

**Comment by max-nextcloud** [comment]
Hi y'all.

Thanks for your input on this issue. I'm working on it right now. My implementation aims to follow what text does in general. First of all because of consistency, second also because it's easier. So i am aiming at:
1. collaborative edititing
2. rendering the tables in the editor and editing them rather than the source code
3. preserving the meaning of markdown - but not necessarily the exact code.

With respect to 2. I hope to look into a markdown view at some point but that would be for all of text - not just for tables.

Regarding 3. I am investigating ways to minimize the diff when editing markdown files in text. For now we read the entire file, convert it to html, load it into tiptap, fetch the html from tiptap and then store that as markdown. We'll have to keep track of the initial markdown or at least some of its properties to preserve it. This is best discussed in #593.

**Comment by juliushaertl** [comment]
Quick early design feedback form the spontaneous design co-working:

- for the header, only the bottom border should be darker, possibly no background color
- there should be a border radius on the table
- some margin at the bottom would help to separate from the next block
- keyboard actions
  - enter in the last row should create a new row
  - tab navigation should work within cells
- when the table is focussed, buttons could be shown for adding/removing rows/columns outside of the table
- https://materialdesignicons.com/icon/table might be a nicer icon
